### PR TITLE
Added 'ChannelDescriptionChangedEvent'

### LINKED ISSUE #1753: What event is published when a binding is changing dynamically a list of state options
When a binding is changing dynamically a list of state options (using a DynamicStateDescriptionProvider), the result of my tests (listening to topic `openhab/items/*/*`) is that there is no item event published on the event bus to inform of this change.

Is there at least an event published ,maybe on another topic ?

I would have expected an event ItemUpdatedEvent with the new state description.

@cweitkamp : I am sure you are aware of that.

### PR DESCRIPTION
**Pull request by cweitkamp**
- Added `ChannelDescriptionChangedEvent`

Problem statement: 

OHC framework supports dynamic `StateDescription`s and dynamic `CommandDescription`s which are set by bindings and may change during runtime. When the state options / command options are changed the UI doesn't update automatically. You need to do a page refresh to pick up the changes. This approach introduces a new event (`ChannelDescriptionChangedEvent`) which UIs can subscribe to and reload related widgets. The event payload contains the `ChannelUID`, a field identifier to give a hint which field has changed and a list of linked item names.

See e.g. https://github.com/openhab/openhab-addons/pull/7396#issuecomment-630773749

Closes #1753 

Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

### CODE REVIEW

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/binding/AbstractDynamicDescriptionProvider.java
Comment (cweitkamp) [inline bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/binding/AbstractDynamicDescriptionProvider.java:33]:
```suggestion
 * The {@link AbstractDynamicDescriptionProvider} provides a base implementation for dynamic description provider.s
```

**Comment by kaikreuzer** [comment]
I so far always considered those descriptions rather static and the "dynamic" providers mainly being a means to have the list defined at runtime and not at compile time (as with the xml files) - thus I never expected UIs to react upon changes.

If you think this is worthwhile to have, it should be fine to add it. You should just be clear of the implications: You should convince all UI devs to react to those events sooner or later and this includes Basic UI, the new Default UI as well as Android, iOS and Windows apps...

**Comment by ghys** [comment]
I agree with @cdjackson that it's good to have the info no matter what's done with it.

A few notes however:
- Sitemap rendering UIs (i.e. mobile apps, Windows, Basic UI, etc.) listen to events on a dedicated SSE connection (`/rest/sitemaps/{sitemapname}/{pageid}`), not the "regular" event bus (`/rest/events`), Would this event appear in those?
- End user UIs usually deal with items, and don't care about channels, so even if the event contains the linked items for the channel, maybe it would be more natural to issue events for the affected item's themselves rather than the channel;
- The new "main" UI tries not to listen to the event bus or maintain a local copy of item descriptions for end-user pages (for performance reasons), but only tracks select items' states with `/rest/events/states`. Command options are currently displayed with "action sheets" like this by default:



In this case, since the options are not displayed until the user clicks on the widget, it would be safer just to make a quick call to `/rest/items/{itemName}` to get the up-to-date command options before displaying the actions. I suppose there will be a need for alternative rendering of these command options (e.g. with buttons when their number is limited), then of course it's another story. Nonetheless, maybe it would be worth looking into adding state descriptions/command descriptions and their updates to the [StateDTO](https://github.com/openhab/openhab-core/blob/master/bundles/org.openhab.core.io.rest.sse/src/main/java/org/openhab/core/io/rest/sse/internal/dto/StateDTO.java) the UI receives whenever an item tracked with `/rest/events/states` changes.

**Comment by cweitkamp** [comment]
Thanks for you useful input. The pointer to SSE events was a good hint. It obviously makes a lot of sense to emit the `ChannelDescriptionChangedEvent` via them too. I will look into that and try to enhance my current implementation to add support for it.

**Comment by mueller-ma** [comment]
> Sitemap rendering UIs (i.e. mobile apps, Windows, Basic UI, etc.) listen to events on a dedicated SSE connection (/rest/sitemaps/{sitemapname}/{pageid}), not the "regular" event bus (/rest/events), Would this event appear in those?

Ideally a SSE will be sent to all affected sitemap elements as well.

/cc @maniac103

**Comment by lolodomo** [comment]
> Thanks for you useful input. The pointer to SSE events was a good hint. It obviously makes a lot of sense to emit the `ChannelDescriptionChangedEvent` via them too. I will look into that and try to enhance my current implementation to add support for it.

For the remote openHAB binding, an event on the topic openhab/items/*/xxx would be ideal for me. But if I have to listen to another topic, why not.
I have imagined it could have be done through the existing event ItemUpdatedEvent.

Do you plan to rework and finish this enhancement ? Before OH3 is released ?

**Comment by cweitkamp** [comment]
> Do you plan to rework and finish this enhancement ? Before OH3 is released ?

I added it to the [openHAB 3 issue tracking](https://github.com/orgs/openhab/projects/3). So yes, I am planning to do further work on it.

**Comment by lolodomo** [comment]
Any progress on this enhancement?

**Comment by cweitkamp** [comment]
I rebased the branch onto latest main. From my POV the PR is in a final state and ready for review.

grafik

I have to admit that I was not able to look deeper into SSE connections. Feel free to give me a hand on that.

**Comment by lolodomo** [comment]
Does it mean that any binding state options provider should then be updated to set the reference to ItemChannelLinkRegistry?

**Comment by lolodomo** [comment]
I think we should publish an event to the item topic too.

**Comment by cweitkamp** [comment]
> Does it mean that any binding state options provider should then be updated to set the references to ItemChannelLinkRegistry and EventPublisher?

AFAIR yes. At least when we want to use constructor injection. I did not test if field injection is working when using abstract base classes. Method injection did not work for it in the past.

**Comment by lolodomo** [comment]
Isn't it possible to provide the new state description in the new event? This would avoid additional request(s) to get it.

**Comment by cweitkamp** [comment]
Sounds like a reasonable idea.

**Comment by lolodomo** [comment]
I think there is no reason to force 
 channelTypeI18nLocalizationService and eventPublisher as not null by default in the abstract class. In fact, they will be null if a binding provider does not set them.

**Comment by cweitkamp** [comment]
@lolodomo Thanks for placing your ideas. I tested them one by one.

I am afraid we cannot use field injection or method injection in abstract classes. It looks like OSGi framework only considers `@Reference` annotations in the final class annotated with `@Component`. We have to adopt all existing implementations to support this new event.

> Isn't it possible to provide the new state description in the new event?

No, that is not possible. It is not available and we do not have access to the full state / command description in the provider classes. We only can add a partial state description which contains only the changed field.

**Comment by lolodomo** [comment]
You could at least provide the new options in the event (or the new pattern).

**Comment by cweitkamp** [comment]
That is possible. Looking into other events for updating an entity (e.g. Thing or Item) we publish the old entity beside the new one.

**Comment by lolodomo** [comment]
Yes, that's true. Is there any advantage to provide the old list of options in addition to the new list of options, I am not sure. But if you want to provide the new AND old list of options, that's fine for me.

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/events/ChannelDescriptionChangedEvent.java
Comment (lolodomo) [inline bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/events/ChannelDescriptionChangedEvent.java:36]: Is ALL necessary (not used I believe) ?

File: bundles/org.openhab.core.thing/src/test/java/org/openhab/core/thing/binding/BaseDynamicCommandDescriptionProviderTest.java
Comment (lolodomo) [inline bundles/org.openhab.core.thing/src/test/java/org/openhab/core/thing/binding/BaseDynamicCommandDescriptionProviderTest.java:88]: Suggestion: you could additionally checked the linked items and the list of options in the event.

File: bundles/org.openhab.core.thing/src/test/java/org/openhab/core/thing/binding/BaseDynamicStateDescriptionProviderTest.java
Comment (lolodomo) [inline bundles/org.openhab.core.thing/src/test/java/org/openhab/core/thing/binding/BaseDynamicStateDescriptionProviderTest.java:106]: Suggestion: you could additionally checked the linked items and the list of options in the event.

File: bundles/org.openhab.core.thing/src/test/java/org/openhab/core/thing/binding/BaseDynamicStateDescriptionProviderTest.java
Comment (lolodomo) [inline bundles/org.openhab.core.thing/src/test/java/org/openhab/core/thing/binding/BaseDynamicStateDescriptionProviderTest.java:88]: Suggestion: you could additionally checked the linked items and the pattern in the event.

**Review by lolodomo — CHANGES_REQUESTED** [review]
Few remarks and suggestions

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/binding/AbstractDynamicDescriptionProvider.java
Comment (cweitkamp) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/binding/AbstractDynamicDescriptionProvider.java]: Done.

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/events/ChannelDescriptionChangedEvent.java
Comment (cweitkamp) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/events/ChannelDescriptionChangedEvent.java:36]: It will be needed if a binding publishes the whole `StateDEscription` (e.g. https://github.com/openhab/openhab-addons/pull/10718). Note: I have to change PRs in openHAB add-ons according the new API.

File: bundles/org.openhab.core.thing/src/test/java/org/openhab/core/thing/binding/BaseDynamicCommandDescriptionProviderTest.java
Comment (cweitkamp) [reply bundles/org.openhab.core.thing/src/test/java/org/openhab/core/thing/binding/BaseDynamicCommandDescriptionProviderTest.java:88]: The `ThingEventFactoryTest` class is responsible for testing the correct payloads.

**Review by lolodomo — APPROVED** [review]
LGTW

**Comment by lolodomo** [comment]
Delay is now short before 3.1 is released. Can this PR be merged ASAP. We have then to enhance 33 bindings (some PRs are already submitted waiting for the merge of this one).

**Review by lolodomo — APPROVED** [review]

**Review by kaikreuzer — APPROVED** [review]
lgtm, thanks!

**Comment by lolodomo** [comment]
Very good.
Can you please trigger a new core build so that we can have this enhancement in the next snapshot ? I need it ot test my changes in the remoteopenhab binding.

Regarding impact on UI (sitemap), I know how to handle that. First a little change is needed in the IO REST sitemap bundle: my idea is to add a boolean field to the widget SSE event to indicate if the item description was changed. This will have no impact on UIs not reading this field. In Basic UI, I will read this field and when set to true, I will trigger the reload of the current page. Nothing very difficult I believe, I just need to validate that it works as I imagine.

**Comment by kaikreuzer** [comment]
@lolodomo I had triggered a build right after merging this PR: https://ci.openhab.org/job/openHAB-Core/793/

**Comment by lolodomo** [comment]
Perfect. 

Note that we have already 3 PRs for addons ready to be merged that enhance bindings with this new feature. 
@cweitkamp is working on the PR for the hue  binding. 
I will try to create another PR covering the remaining bindings this evening.

**Comment by lolodomo** [comment]
@cweitkamp : what is smarthome/j  ?

**Comment by cweitkamp** [comment]
A third party repository for OHC based add-ons.

**Comment by lolodomo** [comment]
I identified 3 bindings implementing a state description provider extending BaseDynamicStateDescriptionProvider in an unusual way: spotify, bticinosmarther and heos. These bindings require a special look. All the other (except hue) are now updated.

And for the 3 bindings implementing a command description provider extending BaseDynamicCommandDescriptionProvider, only the nanoleaf binding is not yet enhanced.

**Comment by lolodomo** [comment]
@cweitkamp : I just discovered we have many bindings directly implementing DynamicStateDescriptionProvider. Those ones will not benefit this enhancement?

**Comment by cweitkamp** [comment]
Yes, that is correct. They have to take care on their own.

**Pull request merged**
